force-app/main/default/aura/expressionBuilder/expressionBuilderController.js

```diff
@@ -0,0 +1,45 @@
+({
+    doInit: function (cmp, event, helper) {
+
+
+
+        //TODO: criteria parsing from string
+
+
+        cmp.set("v.criterias",
+            [{ criteria: "$Record.Contact.LastModifiedDate &gt; (TODAY()-30)", condition: "AND" },
+            { criteria: "$Record.Contact.LastModifiedDate &lt; (TODAY())" }]);
+    },
+
+    handleCriteriaDelete: function (cmp, event, helper) {
+        var criterias = cmp.get("v.criterias");
+        var index = event.getParam("index");
+
+        //if we deleted last element, we should clear condition of previous element
+        if (criterias.length != 1 && criterias.length == index + 1) {
+            criterias[index - 1].condition = "";
+        }
+
+        criterias.splice(index, 1);
+        cmp.set("v.criterias", criterias);
+
+        helper.notifyExpressionUpdate(cmp, event, helper);
+    },
+
+    handleCriteriaAdd: function (cmp, event, helper) {
```

Comment:
I can't figure out what's going on in this function. When a criterion gets added, it always gets added to the bottom of the list of controls. you dynamically create a new criterion component and add it to the criteria's criterionList attribute.